lib/web/desktop.go

```diff
@@ -384,21 +386,48 @@ func (h *Handler) performSessionMFACeremony(
 		span.End()
 	}()
 
+	// channelID is used by the front end to differentiate between separate ongoing SSO challenges.
+	var channelID string
+
 	mfaCeremony := &mfa.Ceremony{
-		PromptConstructor: func(po ...mfa.PromptOpt) mfa.Prompt {
+		CreateAuthenticateChallenge: sctx.cfg.RootClient.CreateAuthenticateChallenge,
+		SSOMFACeremonyConstructor: func(_ context.Context) (mfa.SSOMFACeremony, error) {
+			id, err := uuid.NewRandom()
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			channelID = id.String()
+
+			u, err := url.Parse(sso.WebMFARedirect)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			u.RawQuery = url.Values{"channel_id": {channelID}}.Encode()
+			return &sso.MFACeremony{
+				ClientCallbackURL: u.String(),
+				ProxyAddress:      h.PublicProxyAddr(),
+			}, nil
+		},
+		PromptConstructor: func(...mfa.PromptOpt) mfa.Prompt {
```

Comment:
Do `SSOMFACeremonyConstructor` and `PromptConstructor` run on the same goroutine?

Reply:
Yes. If the concern is that the channelID is set in `SSOMFACeremonyConstructor`, I believe I just did that so we aren't generating unused UUID's when SSO MFA is not an option. Seems like an unnecessary and bug prone optimization in hindsight.